### ISSUE TITLE
`accounts.email.accounts.<name>.userName` must be defined when `programs.git` is enabled

### ISSUE THREAD

**Issue by qaristote** [nix-community/home-manager#3712]
I have defined an email account with 
```nix
accounts.email.accounts.personal = { ... };
```
but didn't specify a `userName` as for IMAP and SMTP it's the same as the address.
I have also enabled git with
```nix
programs.git.enable = true;
```
When building my configuration, I got
```
error: A definition for option `home-manager.users.qaristote.programs.git.iniContent."sendemail.personal".smtpUser' is not of type `string or boolean or signed integer or list of (string or boolean or signed integer) or attribute set of (string or boolean or signed integer or list of (string or boolean or signed integer))'. Definition values:
```

I believe the lines at fault are the following:
https://github.com/nix-community/home-manager/blob/ab7c8f4a8427bfcaf01a46bab974298cc27bc1f5/modules/programs/git.nix#L435-L464

and that a simple fix would be to replace 
```     
hasSmtp = name: account: account.smtp != null; 
```
with
```
hasSmtp = name: account: account.smtp != null && (account.mstmp.enable || account.userName != null);
``` 
but I'll let you see that for yourself :)

**Comment by ncfavier** [comment]
`if (userName != null) then userName else address` should work, do you want to open a PR?

**Comment by qaristote** [comment]
Yeah my bad. Another solution would be to set the default value of `userName` to the value of `address`. I'd rather let the maintainers choose between these solutions and create the PR accordingly !

**Comment by rycee** [comment]
I don't recall why the `userName` option has type `types.nullOr types.str`. Seems to me that it should always be given. Then you would have gotten a better error message. It doesn't seem so nice to just assume that the user name is the same as the email address, better make it clear that it is needed.

**Comment by haennes** [comment]
> I don't recall why the `userName` option has type `types.nullOr types.str`. Seems to me that it should always be given. Then you would have gotten a better error message. It doesn't seem so nice to just assume that the user name is the same as the email address, better make it clear that it is needed.

Should i make a PR that changes the type to just be `types.str` ?

**Comment by teto** [comment]
@haennes go ahead

**Comment by haennes** [comment]
@rycee Finally came around to change userName to require a string. However this breaks quite a lot of unit tests, that just enable git, which will probably lead to a lot of  broken configs.  Should  I continue implementing this?

PS: This is my first time contributing to home-manger / such a big project. I am trying my best :)
